src/mpi/request/wait.c

```diff
@@ -92,10 +92,15 @@ int MPIR_Wait_impl(MPI_Request * request, MPI_Status * status)
         MPID_Progress_end(&progress_state);
     }
 
-    mpi_errno = MPIR_Request_complete(request, request_ptr, status, &active_flag);
+    mpi_errno = MPIR_Request_completion_processing(request_ptr, status, &active_flag);
     if (mpi_errno)
         MPIR_ERR_POP(mpi_errno);
 
+    if (!MPIR_Request_is_persistent(request_ptr)) {
```

Comment:
We cannot check error first, since in the case of error it will return without freeing the request and so leaking it.

Reply:
Oh, you are right.  My bad.  Sorry for breaking that.